// lunar-typescript/dist/lib/FotoFestival.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.FotoFestival = void 0;
var FotoFestival = (function () {
    function FotoFestival(name, result, everyMonth, remark) {
        if (result === void 0) { result = ''; }
        if (everyMonth === void 0) { everyMonth = false; }
        if (remark === void 0) { remark = ''; }
        this._name = name;
        this._result = result ? result : '';
        this._everyMonth = everyMonth;
        this._remark = remark;
    }
    FotoFestival.prototype.getName = function () {
        return this._name;
    };
    FotoFestival.prototype.getResult = function () {
        return this._result;
    };
    FotoFestival.prototype.isEveryMonth = function () {
        return this._everyMonth;
    };
    FotoFestival.prototype.getRemark = function () {
        return this._remark;
    };
    FotoFestival.prototype.toString = function () {
        return this._name;
    };
    FotoFestival.prototype.toFullString = function () {
        var l = [this._name];
        if (this._result) {
            l.push(this._result);
        }
        if (this._remark) {
            l.push(this._remark);
        }
        return l.join(' ');
    };
    return FotoFestival;
}());
exports.FotoFestival = FotoFestival;
