// lunar-typescript/dist/lib/JieQi.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.JieQi = void 0;
var Lunar_1 = require("./Lunar");
var JieQi = (function () {
    function JieQi(name, solar) {
        var jie = false, qi = false, i, j;
        for (i = 0, j = Lunar_1.Lunar.JIE_QI.length; i < j; i++) {
            if (Lunar_1.Lunar.JIE_QI[i] === name) {
                if (i % 2 == 0) {
                    qi = true;
                }
                else {
                    jie = true;
                }
                break;
            }
        }
        this._name = name;
        this._solar = solar;
        this._jie = jie;
        this._qi = qi;
    }
    JieQi.prototype.getName = function () {
        return this._name;
    };
    JieQi.prototype.getSolar = function () {
        return this._solar;
    };
    JieQi.prototype.setName = function (name) {
        this._name = name;
    };
    JieQi.prototype.setSolar = function (solar) {
        this._solar = solar;
    };
    JieQi.prototype.isJie = function () {
        return this._jie;
    };
    JieQi.prototype.isQi = function () {
        return this._qi;
    };
    JieQi.prototype.toString = function () {
        return this.getName();
    };
    return JieQi;
}());
exports.JieQi = JieQi;
