// lunar-typescript/dist/lib/LiuNian.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.LiuNian = void 0;
var LunarUtil_1 = require("./LunarUtil");
var LiuYue_1 = require("./LiuYue");
var LiuNian = (function () {
    function LiuNian(daYun, index) {
        this._year = daYun.getStartYear() + index;
        this._age = daYun.getStartAge() + index;
        this._index = index;
        this._daYun = daYun;
        this._lunar = daYun.getLunar();
    }
    LiuNian.prototype.getYear = function () {
        return this._year;
    };
    LiuNian.prototype.getAge = function () {
        return this._age;
    };
    LiuNian.prototype.getIndex = function () {
        return this._index;
    };
    LiuNian.prototype.getLunar = function () {
        return this._lunar;
    };
    LiuNian.prototype.getGanZhi = function () {
        var offset = LunarUtil_1.LunarUtil.getJiaZiIndex(this._lunar.getJieQiTable().get('立春').getLunar().getYearInGanZhiExact()) + this._index;
        if (this._daYun.getIndex() > 0) {
            offset += this._daYun.getStartAge() - 1;
        }
        offset %= LunarUtil_1.LunarUtil.JIA_ZI.length;
        return LunarUtil_1.LunarUtil.JIA_ZI[offset];
    };
    LiuNian.prototype.getXun = function () {
        return LunarUtil_1.LunarUtil.getXun(this.getGanZhi());
    };
    LiuNian.prototype.getXunKong = function () {
        return LunarUtil_1.LunarUtil.getXunKong(this.getGanZhi());
    };
    LiuNian.prototype.getLiuYue = function () {
        var l = [];
        for (var i = 0; i < 12; i++) {
            l.push(new LiuYue_1.LiuYue(this, i));
        }
        return l;
    };
    return LiuNian;
}());
exports.LiuNian = LiuNian;
