// lunar-typescript/dist/lib/SolarMonth.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SolarMonth = void 0;
var SolarUtil_1 = require("./SolarUtil");
var Solar_1 = require("./Solar");
var SolarWeek_1 = require("./SolarWeek");
var ExactDate_1 = require("./ExactDate");
var SolarMonth = (function () {
    function SolarMonth(year, month) {
        this._year = year;
        this._month = month;
        this._calendar = ExactDate_1.ExactDate.fromYmd(year, month, 1);
    }
    SolarMonth.fromYm = function (year, month) {
        return new SolarMonth(year, month);
    };
    SolarMonth.fromDate = function (date) {
        return SolarMonth.fromYm(date.getFullYear(), date.getMonth() + 1);
    };
    SolarMonth.prototype.getYear = function () {
        return this._year;
    };
    SolarMonth.prototype.getMonth = function () {
        return this._month;
    };
    SolarMonth.prototype.next = function (months) {
        var date = ExactDate_1.ExactDate.fromYmd(this._year, this._month, 1);
        date.setMonth(date.getMonth() + months);
        return SolarMonth.fromDate(date);
    };
    SolarMonth.prototype.getDays = function () {
        var l = [];
        var d = Solar_1.Solar.fromYmd(this._year, this._month, 1);
        l.push(d);
        var days = SolarUtil_1.SolarUtil.getDaysOfMonth(this._year, this._month);
        for (var i = 1; i < days; i++) {
            l.push(d.next(i));
        }
        return l;
    };
    SolarMonth.prototype.getWeeks = function (start) {
        var l = [];
        var week = SolarWeek_1.SolarWeek.fromYmd(this._year, this._month, 1, start);
        while (true) {
            l.push(week);
            week = week.next(1, false);
            var firstDay = week.getFirstDay();
            if (firstDay.getYear() > this._year || firstDay.getMonth() > this._month) {
                break;
            }
        }
        return l;
    };
    SolarMonth.prototype.toString = function () {
        return "".concat(this.getYear(), "-").concat(this.getMonth());
    };
    SolarMonth.prototype.toFullString = function () {
        return "".concat(this.getYear(), "\u5E74").concat(this.getMonth(), "\u6708");
    };
    return SolarMonth;
}());
exports.SolarMonth = SolarMonth;
